### PR TITLE
Separable module configuration -- without the bugs this time

### LINKED ISSUE #23663: AttributeError: 'CommentedSeq' object has no attribute 'get'
I'm trying to install a new package after pulling spack to a recent version; and the install seems to be failing due to an error while building modulefiles deep in the spack config settings.

### Error Message

```console
$ spack -vd install fortrilinos ^trilinos/fi45zpy
# snip
==> [2021-05-16-09:23:02.993000] Installing /var/folders/gy/mrg1ffts2h945qj9k29s1l1dvvmbqb/T/s3j/spack-stage/spack-stage-fortrilinos-2.0.0-xdulgmsdgorsfvtvr73vubopurrs2hn5/spack-src/example/test-installation to /rnsdhpc/code/spack/opt/spack/apple-clang/fortrilinos/xdulgms/.spack/test/example/test-installation
==> [2021-05-16-09:23:03.317370] Error: AttributeError: 'CommentedSeq' object has no attribute 'get'

/rnsdhpc/code/spack/lib/spack/spack/build_environment.py:930, in _setup_pkg_and_run:
        927        tb_string = traceback.format_exc()
        928
        929        # build up some context from the offending package so we can
  >>    930        # show that, too.
        931        package_context = get_package_context(tb)
        932
        933        logfile = None

See build log for details:
  /var/folders/gy/mrg1ffts2h945qj9k29s1l1dvvmbqb/T/s3j/spack-stage/spack-stage-fortrilinos-2.0.0-xdulgmsdgorsfvtvr73vubopurrs2hn5/spack-build-out.txt

Traceback (most recent call last):
  File "/rnsdhpc/code/spack/lib/spack/spack/build_environment.py", line 914, in _setup_pkg_and_run
    return_value = function(pkg, kwargs)
  File "/rnsdhpc/code/spack/lib/spack/spack/installer.py", line 1792, in build_process
    spack.hooks.post_install(pkg.spec)
  File "/rnsdhpc/code/spack/lib/spack/spack/hooks/__init__.py", line 75, in __call__
    hook(*args, **kwargs)
  File "/rnsdhpc/code/spack/lib/spack/spack/hooks/module_file_generation.py", line 38, in post_install
    _for_each_enabled(spec, 'write')
  File "/rnsdhpc/code/spack/lib/spack/spack/hooks/module_file_generation.py", line 15, in _for_each_enabled
    enabled = spack.config.get('modules:%s:enable' % name)
  File "/rnsdhpc/code/spack/lib/spack/spack/config.py", line 899, in get
    return config.get(path, default, scope)
  File "/rnsdhpc/code/spack/lib/spack/spack/config.py", line 652, in get
    value = value.get(key, default)
AttributeError: 'CommentedSeq' object has no attribute 'get'
==> [2021-05-16-09:23:03.317914] Flagging fortrilinos-2.0.0-xdulgmsdgorsfvtvr73vubopurrs2hn5 as failed: AttributeError: 'CommentedSeq' object has no attribute 'get'
==> [2021-05-16-09:23:03.333998] ChildError: AttributeError: 'CommentedSeq' object has no attribute 'get'
```

### Information on your system

* **Spack:** 0.16.1-2661-2202ce27fa
* **Python:** 3.8.5
* **Platform:** darwin-bigsur-skylake
* **Concretizer:** clingo

### Additional information

- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have searched the issues of this repo and believe this is not a duplicate
- [x] I have run the failing commands in debug mode and reported the output

The modules config looks a little funky, apparently 
```console
$ spack config get modules
modules:
  default:
    lmod:
      core_compilers:
      - apple-clang@12.0.5
      - gcc@10.2.0
      hierarchy:
      - mpi
    prefix_inspections:
      lib:
      - DYLD_FALLBACK_LIBRARY_PATH
      lib64:
      - DYLD_FALLBACK_LIBRARY_PATH
    roots:
      tcl: $spack/share/spack/modules
      lmod: $spack/share/spack/lmod
    enable:
    - tcl

    # Default configurations if lmod is enabled
  'enable:':
  - lmod
  prefix_inspections:
    bin:
    - PATH
    man:
    - MANPATH
    share/man:
    - MANPATH
    share/aclocal:
    - ACLOCAL_PATH
    'include:': []
    lib:
    - LIBRARY_PATH
    lib64:
    - LIBRARY_PATH
    lib/pkgconfig:
    - PKG_CONFIG_PATH
    lib64/pkgconfig:
    - PKG_CONFIG_PATH
    share/pkgconfig:
    - PKG_CONFIG_PATH
    ? ''
    : - CMAKE_PREFIX_PATH

  # These are configurations for the module set named "default"
  lmod:
    core_compilers:
    - apple-clang@12.0.0
    hash_length: 4
    'hierarchy:': []
    ^python:
      autoload: direct
    all:
      suffixes:
        ^python@2: py2
        +cuda: cuda
        +debug: debug
        +mpi: mpi
        +multithreaded: mt
        +openmp: mt
        +threads: mt
        ~shared: static
        cxxstd=98: c++98
        cxxstd=14: c++14
        cxxstd=17: c++17
      environment:
        set:
          ${PACKAGE}_ROOT: ${PREFIX}
```

### LINKED ISSUE #23665: environment w/ view: spack env activate: PATH not set right on RHEL8
Here is a case where `spack env activate` does not set `PATH` (and other environment variables) on RHEL 8

Using:
* `spack@develop` (379a1fd23bf464717b93a09fde68cf6469c2cbe7 from `Sat May 15 13:48:00 2021 -0700`)
* RHEL8 via container image `ecpe4s/rhel8-runner-x86_64:2021-05-16`

spack.yaml:
```
 spack:
  view: /bootstrap/view
  concretization: separately

  config:
    install_tree:
      root: /bootstrap/install

  packages:
    all:
      target: [x86_64]

  specs:
  - patchelf
```

```
$> spack -e . install --no-cache
==> Concretized patchelf
 -   hfkgaau  patchelf@0.12%gcc@8.3.1 patches=a155f233b228f02d7886e304cb13898d93801b52f351e098c2cc0719697ec9d0 arch=linux-rhel8-x86_64

==> Installing environment /
==> Installing patchelf-0.12-hfkgaauno6l77h5ds4hrj4xgpqalfmei
==> Fetching https://spack-llnl-mirror.s3-us-west-2.amazonaws.com/_source-cache/archive/69/699a31cf52211cf5ad6e35a8801eb637bc7f3c43117140426400d67b7babd792.tar.bz2
############################################################################################################################################### 100.0%
==> Fetching https://spack-llnl-mirror.s3-us-west-2.amazonaws.com/_source-cache/archive/a1/a155f233b228f02d7886e304cb13898d93801b52f351e098c2cc0719697ec9d0
############################################################################################################################################### 100.0%
==> Applied patch https://github.com/NixOS/patchelf/pull/230.patch
==> patchelf: Executing phase: 'autoreconf'
==> patchelf: Executing phase: 'configure'
==> patchelf: Executing phase: 'build'
==> patchelf: Executing phase: 'install'
==> patchelf: Successfully installed patchelf-0.12-hfkgaauno6l77h5ds4hrj4xgpqalfmei
  Fetch: 0.37s.  Build: 8.12s.  Total: 8.49s.
[+] /bootstrap/install/linux-rhel8-x86_64/gcc-8.3.1/patchelf-0.12-hfkgaauno6l77h5ds4hrj4xgpqalfmei
==> Updating view at /bootstrap/view

$> ls -l /bootstrap/view/bin
patchelf

$> /bootstrap/view/bin/patchelf --version
patchelf 0.12.20200827.8d3a16e

$> spack env activate -d .

$> which patchelf
/usr/bin/which: no patchelf in (/spack/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin)

$> spack env activate -d . --sh
export SPACK_ENV=/;
alias despacktivate='spack env deactivate';
```

This exact same environment can be activated on RHEL 7, Ubuntu 18.04, and Ubuntu 20.04 and `PATH` gets set correctly there. Just not here.

### LINKED ISSUE #23701: `spack load` broken by #22588
### Before

```console
$ git checkout cefbe48c89209dc3df654795644973b1885cdea4~1
$ spack load --first --sh python
# modifies the following variables:
SPACK_LOADED_HASHES
CPATH
MANPATH
PKG_CONFIG_PATH
PATH
ACLOCAL_PATH
DYLD_FALLBACK_LIBRARY_PATH
CMAKE_PREFIX_PATH
```

### After

```console
$ git checkout cefbe48c89209dc3df654795644973b1885cdea4
$ spack load --first --sh python
# modifies the following environment variables:
DYLD_FALLBACK_LIBRARY_PATH
CPATH
SPACK_LOADED_HASHES
```

My `modules.yaml` looks like:
```yaml
modules:
  enable:: []
```
but even without this file I can reproduce the bug. Any idea what could be causing this?

* **Spack:** 0.16.1-2651-fc392d2f56
* **Python:** 3.8.7
* **Platform:** darwin-catalina-ivybridge
* **Concretizer:** clingo

### PR DESCRIPTION
**Pull request by becker33**
Reimplements #22588 

fixes #23701 
fixes #23663
fixes #23665 

Description for 22588 reproduced below:

Currently, module configurations are inconsistent because modulefiles are generated with the configs for the active environment, but are shared among all environments (and spack outside any environment).

This PR fixes that by allowing Spack environments (or other spack config scopes) to define additional sets of modules to generate. Each set of modules can enable either lmod or tcl modules, and contains all of the previously available module configuration. The user defines the name of each module set -- the set configured in Spack by default is named "default", and is the one returned by module manipulation commands in the absence of user intervention.

As part of this change, the module roots configuration moved from the config section to inside each module configuration.

Additionally, it adds a feature that the modulefiles for an environment can be configured to be relative to an environment view rather than the underlying prefix. This will not be enabled by default, as it should only be enabled within an environment and for non-default views constructed with separate projections per-spec.

TODO:
- [x] Documentation

### CODE REVIEW

**Comment by adamjstewart** [comment]
I can confirm that this fixes #23701

**Comment by tgamblin** [comment]
@becker33: in addition to the docs -- can you try for 100% patch coverage on this?  If it's way too tedious let's skip it but in light of the issues I think it'd be good to get the coverage up.

**Comment by becker33** [comment]
@tgamblin my instinct is that will be incredibly difficult, but I will look into it when I solve the reported issues.

**Comment by tgamblin** [comment]
@becker33: at the very least I think the reported issues should have regression tests -- as you say let's see where the coverage is with that.

**Comment by eugeneswalker** [comment]
FYI Fixes https://github.com/spack/spack/issues/23665

**Comment by becker33** [comment]
Kicking tests

**Review by tgamblin — APPROVED** [review]

File: etc/spack/defaults/linux/modules.yaml [hunk 1]
```diff
@@ -14,8 +14,9 @@
 #   ~/.spack/modules.yaml
```
Comment (adamjstewart) [inline etc/spack/defaults/linux/modules.yaml:14]: Why was `darwin/modules.yaml` not updated?

**Pull request merged**